File: Arduino/RoboticArm/src/kinematics.cpp
#include "kinematics.h"
#include "config.h"
#include "control.h"
#include <math.h>

void inverse_kinematics(float x, float y, float z) {
    z -= h;
    float theta = (x != 0) ? atan2(y, x) : (y > 0 ? 1 : -1) * PI/2;
    x = sqrt(x*x + y*y);
    float goal_mu_radians = goal_mu * DEG_TO_RAD;

    x +=  -L3*cos(goal_mu_radians) - L4*sin(goal_mu_radians);
    z +=  -L3*sin(goal_mu_radians) + L4*cos(goal_mu_radians);

    float r = sqrt(x*x + z*z);
    float a = acos((-L2*L2 + L1*L1 + r*r)/(2*L1*r));
    float b = acos((L1*L1 + L2*L2 - r*r)/(2*L1*L2));
    float atn = atan2(z, x);

    calculated_inverse[0] = theta * RAD_TO_DEG;
    calculated_inverse[1] = (PI - a - b - atn) * RAD_TO_DEG;
    calculated_inverse[2] = (PI/2 - a - atn) * RAD_TO_DEG;
}

void direct_kinematics(float theta, float alpha, float beta) {
    // Convert degrees to radians
    theta = theta * DEG_TO_RAD;
    alpha = alpha * DEG_TO_RAD;
    beta = beta * DEG_TO_RAD;
    float gamma = (goal_mu * DEG_TO_RAD) + alpha;

    float q1 = PI/2 - beta;
    float q2 = -alpha;
    float q3 = gamma - alpha;

    float x_plane = (
        L1 * cos(q1) +
        L2 * cos(q2) +
        L3 * cos(q3) + L4 * sin(q3)
    );
    float z_plane = (
        L1 * sin(q1) +
        L2 * sin(q2) +
        L3 * sin(q3) - L4 * cos(q3)
    );

    calculated_direct[0] = x_plane * cos(theta);
    calculated_direct[1] = x_plane * sin(theta);
    calculated_direct[2] = z_plane + h;
}

float mu_to_gamma(float mu) {
    return mu + current_angle[1];
}

float gamma_to_mu(float gamma) {
    return gamma - current_angle[1];
}

File: GUI/kinematics.py
import numpy as np

def inverse_kinematics(x_abs, y_abs, z_abs, l1=250, l2=200, l3=150, mu=0):
    theta = np.arctan2(y_abs, x_abs) if x_abs !=0 else (1 if y_abs>0 else -1) * np.pi/2
    x_abs = np.sqrt(x_abs**2+y_abs**2)
    x = x_abs - l3*np.cos(mu)
    z = z_abs - l3*np.sin(mu)
    r = np.sqrt(x**2 + z**2)
    a = np.arccos((-l2**2 + l1**2 + r**2)/(2*l1*r))
    b = np.arccos((l1**2 + l2**2 - r**2)/(2*l1*l2))
    atn = np.arctan2(z, x)

    return [
        theta, #theta
        np.pi - a - b - atn, #alpha
        np.pi/2 - a - atn, #beta
        mu + np.pi - a - b - atn #gamma
    ]

def direct_kinematics(theta, alpha, beta, gamma, l1=250, l2=200, l3=150):
    q1 = np.pi/2 - beta
    q2 = -alpha
    q3 = gamma - alpha

    x_plane = (
        l1 * np.cos(q1) +
        l2 * np.cos(q2) +
        l3 * np.cos(q3)
    )
    z_plane = (
        l1 * np.sin(q1) +
        l2 * np.sin(q2) +
        l3 * np.sin(q3)
    )

    x = x_plane * np.cos(theta)
    y = x_plane * np.sin(theta)
    z = z_plane
    return np.array([x, y, z])

#test
if __name__ == "__main__":
    x_target = 200
    y_target = 100
    z_target = 100
    mu = 0  # End effector orientation

    angles = inverse_kinematics(x_target, y_target, z_target, mu=mu)
    print("Inverse Kinematics Angles (radians):\n", *angles)

    position = direct_kinematics(*angles)
    print("Direct Kinematics Position:\n", position)